docs/stateful.rst

```diff
@@ -255,124 +247,74 @@ invariants, you should store relevant data on the instance instead.
 Generic state machines
 ----------------------
 
-The class GenericStateMachine is the underlying machinery of stateful testing
-in Hypothesis. In execution it looks much like the RuleBasedStateMachine but
-it allows the set of steps available to depend in essentially arbitrary
-ways on what has happened so far. For example, if you wanted to
-use Hypothesis to test a game, it could choose each step in the machine based
-on the game to date and the set of actions the game program is telling it it
-has available.
+The class ``GenericStateMachine`` is the underlying machinery of stateful testing
+in Hypothesis. Chances are you will want to use the rule based stateful testing
+for most things, but the generic state machine functionality can be useful e.g. if
+you want to test things where the set of actions to be taken is more closely
+tied to the state of the system you are testing.
 
-It essentially executes the following loop:
+.. module:: hypothesis.stateful
+.. autoclass:: GenericStateMachine
+    :members: steps, execute_step, check_invariants, teardown
```

Comment:
Related to #1193, so feel free to tell me it's out of scope, but we could omit the list of members and simply ensure that all private members start with `_`...